### DISCUSSION TITLE
Conditional check for String match

### DISCUSSION THREAD

**Discussion by techpranav** [antlr/antlr4#4172] [Q&A] [answered]
Hi,

Is there any possibility to do string match in antlr4 grammar.
Below is my requirement;
for statement ;-  delete table Persons;
this query should be equivalent to 
del table Persons;
dele table Persons;
delet table Persons;

but it should not allow deletes table Persons;

It should match first three characters and maximum equal to actual keyword.

Without giving Hardcoded values , is there possibility to implement this?

**Comment by jimidle** [comment, marked as answer]
Lexer rule should be (assuming case insensitivity (note formatted for readability, it does not need to be so many lines really):
```
DELETE: ('D'|d')('E'|'e')('L'|'l')   // DEL always required
                ( 
                    ('E'|'e')    // DELE
                         ( 
                             ('T'|'t')   // DELET
                                     ('E'|'e')?  // DELETE 
                         )? 
                )? 
      ;
```

You will get the DELETE token for each of DEL DELE DELET DELETE

Unless you feel that this is hard coded values? There is no  syntactic sugar for doing this - you need to give the lexer enough information. Trivial to write a short program to write the lexer rules for you though.

**Reply by techpranav** [reply]
Do I need to write grammar like the above for each keyword?

**Reply by KvanTTT** [reply]
BTW, with enabled [`caseInsensitive`](https://github.com/antlr/antlr4/blob/master/doc/options.md#caseinsensitive) option it looks simpler:

```g4
lexer grammar L;
options { caseInsensitive = true; }

DELETE: 'DEL' ('E' ('T' 'E'?)?)?
```

> Do I need to write grammar like the above for each keyword?

Yes.

**Reply by jimidle** [reply]
The only thing that can happen with the case option though is when you have a language that is sometimes case insensitive and sometimes not. Which is madness, but I have had to deal with it.

**Reply by jimidle** [reply]
> Do I need to write grammar like the above for each keyword?

Yes, but write a small program in python or something that takes your list of keywords and generates the rules once.

**Reply by KvanTTT** [reply]
> The only thing that can happen with the case option though is when you have a language that is sometimes case insensitive and sometimes not. Which is madness, but I have had to deal with it.

@jimidle in this case you can use per-rule [`caseInsensitive`](https://github.com/antlr/antlr4/blob/master/doc/lexer-rules.md#caseinsensitive) option:

```g4
options { caseInsensitive=true; }
STRING options { caseInsensitive=false; } : 'N'? '\'' (~'\'' | '\'\'')* '\''; // lower n is not allowed
```
